precompile/contract_native_minter.go

```diff
@@ -0,0 +1,129 @@
+
+	mintSignature = CalculateFunctionSelector("mint(address,uint256)") // address, amount
+
+	ErrCannotMint = errors.New("non-enabled cannot mint")
+
+	inputLen = common.HashLength + common.HashLength
+)
+
+// ContractNativeMinterConfig wraps [AllowListConfig] and uses it to implement the StatefulPrecompileConfig
+// interface while adding in the contract deployer specific precompile address.
+type ContractNativeMinterConfig struct {
+	AllowListConfig
+}
+
+// Address returns the address of the contract deployer allow list.
+func (c *ContractNativeMinterConfig) Address() common.Address {
+	return ContractNativeMinterAddress
+}
+
+// Configure configures [state] with the desired admins based on [c].
+func (c *ContractNativeMinterConfig) Configure(state StateDB) {
+	c.AllowListConfig.Configure(state, ContractNativeMinterAddress)
+}
+
+// Contract returns the singleton stateful precompiled contract to be used for the allow list.
+func (c *ContractNativeMinterConfig) Contract() StatefulPrecompiledContract {
+	return ContractNativeMinterPrecompile
+}
+
+// GetContractNativeMinterStatus returns the role of [address] for the minter list.
+func GetContractNativeMinterStatus(stateDB StateDB, address common.Address) AllowListRole {
+	return getAllowListStatus(stateDB, ContractNativeMinterAddress, address)
+}
+
+// SetContractNativeMinterStatus sets the permissions of [address] to [role] for the
+// contract deployer allow list.
+// assumes [role] has already been verified as valid.
+func SetContractNativeMinterStatus(stateDB StateDB, address common.Address, role AllowListRole) {
+	setAllowListRole(stateDB, ContractNativeMinterAddress, address, role)
+}
+
+// PackMintInput packs [address] and [amount] into the appropriate arguments for minting operation.
+func PackMintInput(address common.Address, amount *big.Int) ([]byte, error) {
+	// function selector (4 bytes) + input(hash for address + hash for amount)
+	input := make([]byte, 0, selectorLen+inputLen)
+	input = append(input, mintSignature...)
+	input = append(input, address.Hash().Bytes()...)
+	input = append(input, amount.Bytes()...)
+	return input, nil
+}
+
+// UnpackMintInput attempts to unpack [input] into the arguments to the mint precompile
+func UnpackMintInput(input []byte) (common.Address, *big.Int, error) {
+	if len(input) != inputLen {
+		return common.Address{}, nil, fmt.Errorf("invalid input length for minting: %d", len(input))
+	}
+	to := common.BytesToAddress(input[:common.AddressLength])
+	assetAmount := new(big.Int).SetBytes(input[common.AddressLength : common.AddressLength+common.HashLength])
+	return to, assetAmount, nil
+}
+
+// createMint // TODO: add comment
+func createMint(precompileAddr common.Address) RunStatefulPrecompileFunc {
+	return func(evm PrecompileAccessibleState, callerAddr, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
```

Comment:
Not sure if we want to follow the same pattern as for the allow list here to create a minting execution function. I did this with the allow list since it might be re-used for different types of allow lists at different addresses, as well as to use it for writing different roles, but here there is only one native asset to be minted, and I would think that we will only ever deploy it at a single address.

Reply:
done